plugin/pulp/plugin/download.py

```diff
@@ -0,0 +1,122 @@
+DEFAULT_CHUNKSIZE = 1024 * 1024
+
+
+def default_callback_generator(filename=None, size=None, checksums=None):
+    """
+    Creates a closure to write a response to disk, checksum it, and verify
+    its size.
+
+    :param filename: The absolute path the response should be saved to.
+    :param size: The size, in bytes, that the file is expected to be; check is
+                 skipped if this value is not provided.
+    :param checksums: A dictionary of checksums to validate the file with; the
+                      key should be the hash algorithm and the value the
+                      expected hex digest.
+    """
+    # TODO potentially add a way to have this generate the entire checksum set.
+    # Maybe if there's {'algorithm': None}?
+    checksums = checksums or {}
+
+    def callback(self, response):
+        """
+        A callback for the requests_futures library.
+
+        :param resp: The requests response object
+        """
+        hashes = {algorithm: hashlib.new(algorithm) for algorithm in checksums}
+        with open(filename, mode='wb') as fp:
+            chunks = response.iter_content(chunk_size=DEFAULT_CHUNKSIZE)
+            for chunk in chunks:
+                for hash in hashes.values():
+                    hash.update(chunk)
+                fp.write(chunk)
+
+        # Validate that what we downloaded is what the user expects
+        # TODO add more reasonable exceptions
+        if size and size != os.stat(filename).st_size:
+            raise ValueError()
+
+        if checksums != {algorithm: hash.hexdigest() for algorithm, hash in hashes.items()}:
+            raise ValueError()
+
+    return callback
+
+
+class FileFuturesSession(FuturesSession):
+    """
+    This class implements the requests-futures API, with one  small adjuestment. It provides
+    a default callback that writes a file to disk and validates its content. Once it
+    successfully returns, the file is ready to be imported into Pulp.
+
+        >>> future_resp = asynchronous_session.get('https://example.com')
+        >>> filename, resp = future_resp.result()
+        >>> with open(filename) as fd:
+        >>>     django_file = django.core.files.File(fd)
+        >>> artifact = pulp.apps.models.Artifact(file=django_file)
+
+    We should probably provide nice callback options to spit out Artifact objects or even Content
+    objects.
+    """
+
+    def __init__(self, executor=None, max_workers=5, session=None, *args,
+                 **kwargs):
+        super(FileFuturesSession, self).__init__(*args, **kwargs)
+
+    def request(self, *args, **kwargs):
+        """
+        This provides the requests API as well as the requests-futures API.
+
+        The only difference is a few additional keyword arguments are accepted.
+        """
+        callback_generator = kwargs.pop('callback_generator', default_callback_generator)
+        callback_generator_kwargs = kwargs.pop(
```

Comment:
I agree we should leverage the `requests` lib as much as possible and perhaps to the extent of plugin writers using its API directly.  I dissagree that the requests API is clean but it does seem to be popular.  And, there is value in that.  But, like many python libs, is popular mostly because its the best thing available but not because it's necessarily well done.  For example, the `requests.request(method, url, **kwargs)` taking `**kwargs` and documenting what the arguments are is pretty hacky and requires users to look them up in the docs.  I have other criticisms of the requests API but I don't think it useful to enumerate them here.

The `requests_futures` is a thin integration of `requests` and `futures` which sounds good but it's done in an odd way.  **First**, the `FuturesSession`  **IsA** `requests.Session` but it takes a `Session` constructor argument (I have to assume because it's poorly documented) and contains a `session` attribute.  That is the *wonky* part.  **Second**, the `FuturesSession.request()` sanitizes an undocumented keyword parameter named *background_callback* and simply calls `Executor.submit()`.  Then, returns a `Future` which is a deviation from the *requests* API.  Also, overriding a method and returning a different type is bad.  The `Executor.submit()` queues the request in a `Queue` that is configured without a *maxsize*.  So, each call to `request()` results in something stored in memory.  Not good for scalability.  The caller then has to store the returned future(s) in a list (more memory).  Then,  iterate them calling `Future.done()`.  Or, pass in a callback function.